### DISCUSSION TITLE
Table appears again after deletion

### DISCUSSION THREAD

**Discussion by jerebtw** [pocketbase/pocketbase#2551] [Q&A] [answered]
Sometimes when I delete a table and restart pocketbase, the table comes back but without data. (i couldn't find a good way to reproduce the issue)

In the screenshot below you can see the logs after restarting the container:
![image](https://github.com/pocketbase/pocketbase/assets/49494752/f766393c-0ed3-460a-9856-228e68f3f928)

Context:
Running pocketbase inside Docker with the flowing Dockerfile:

```dockerfile
FROM alpine:latest

ARG PB_VERSION=0.16.1

RUN apk add --no-cache \
    unzip \
    ca-certificates

ADD https://github.com/pocketbase/pocketbase/releases/download/v${PB_VERSION}/pocketbase_${PB_VERSION}_linux_amd64.zip /tmp/pb.zip
RUN unzip /tmp/pb.zip -d /pb/

EXPOSE 8080

CMD ["/pb/pocketbase", "serve", "--http=0.0.0.0:8080", "--debug"]
```

<details>
  <summary>cat 1684877244_collections_snapshot.js</summary>

```javascript
migrate((db) => {
  const snapshot = [
    {
      "id": "_pb_users_auth_",
      "created": "2023-01-22 16:04:56.640Z",
      "updated": "2023-05-23 21:26:01.118Z",
      "name": "users",
      "type": "auth",
      "system": false,
      "schema": [
        {
          "system": false,
          "id": "users_name",
          "name": "name",
          "type": "text",
          "required": false,
          "unique": false,
          "options": {
            "min": null,
            "max": null,
            "pattern": ""
          }
        }
      ],
      "indexes": [
        "CREATE INDEX `__pb_users_auth__created_idx` ON `users` (`created`)"
      ],
      "listRule": "id = @request.auth.id",
      "viewRule": "id = @request.auth.id",
      "createRule": "",
      "updateRule": "id = @request.auth.id",
      "deleteRule": "id = @request.auth.id",
      "options": {
        "allowEmailAuth": true,
        "allowOAuth2Auth": true,
        "allowUsernameAuth": true,
        "exceptEmailDomains": null,
        "manageRule": null,
        "minPasswordLength": 8,
        "onlyEmailDomains": null,
        "requireEmail": false
      }
    },
    {
      "id": "n92zrkbyu0n2k2q",
      "created": "2023-05-23 21:26:17.285Z",
      "updated": "2023-05-23 21:26:17.285Z",
      "name": "test",
      "type": "base",
      "system": false,
      "schema": [
        {
          "system": false,
          "id": "ylghsnfi",
          "name": "test",
          "type": "editor",
          "required": false,
          "unique": false,
          "options": {}
        }
      ],
      "indexes": [],
      "listRule": null,
      "viewRule": null,
      "createRule": null,
      "updateRule": null,
      "deleteRule": null,
      "options": {}
    }
  ];

  const collections = snapshot.map((item) => new Collection(item));

  return Dao(db).importCollections(collections, true, null);
}, (db) => {
  return null;
})
```
</details>

<details>
  <summary>cat 1684877258_deleted_test.js</summary>

```javascript
migrate((db) => {
  const dao = new Dao(db);
  const collection = dao.findCollectionByNameOrId("n92zrkbyu0n2k2q");

  return dao.deleteCollection(collection);
}, (db) => {
  const collection = new Collection({
    "id": "n92zrkbyu0n2k2q",
    "created": "2023-05-23 21:26:17.285Z",
    "updated": "2023-05-23 21:26:17.285Z",
    "name": "test",
    "type": "base",
    "system": false,
    "schema": [
      {
        "system": false,
        "id": "ylghsnfi",
        "name": "test",
        "type": "editor",
        "required": false,
        "unique": false,
        "options": {}
      }
    ],
    "indexes": [],
    "listRule": null,
    "viewRule": null,
    "createRule": null,
    "updateRule": null,
    "deleteRule": null,
    "options": {}
  });

  return Dao(db).saveCollection(collection);
})
```
</details>

**Comment by ganigeorgiev** [comment]
Do you have local `pb_migrations` directory? Is it mounted as volume?

**Reply by jerebtw** [reply]
No i only mounted /pb_data
Should it be mounted?

**Reply by ganigeorgiev** [reply]
Yes, either mount it as volume next to the pb_data OR use the ADD/COPY Dockerfile instructions to copy the dir as part of the image.

If it is not mounted, how did you generate the above migrations? Locally with a separate instance outside of the container?

**Reply by jerebtw** [reply]
I connected to the container and used cat if you mean 1684877244_collections_snapshot.js

**Reply by ganigeorgiev** [reply]
I'm confused. If you have only the `pb_data` mounted then on container restart it will load only the data that is available in `pb_data/data.db`. I don't see how the collection will be recreated. Either the delete didn't succeed or something else happened but I'm not sure how to reproduce it.

**Reply by jerebtw** [reply]
I created a test collection with a test field and added a row of random data, then deleted the collection and then restarted the container.
(However, this only sometimes brings the error to light)

I will try to find a reproducible case.

**Reply by ganigeorgiev** [reply]
I think I know how this could have happened. The initial js snapshot migration is created but it is not registered in the `_migrations` table, while the autogenerated delete migration is registered immediately once it is generated (hence there is only a `SELECT` in the screenshot). On Docker restart the container data persists because the container is not destroyed/removed and it will see that there 2 migrations (loaded from some temporary Docker directory) - the snapshot and the delete; the delete was already applied and registered in your `pb_data/data.db` and will be skipped, but the snapshot will be executed since it is not registered yet (at least on the first restart).

I'll try to submit a fix tomorrow.

**Reply by ganigeorgiev** [reply, marked as answer]
Should be fixed in [v0.16.2 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.16.2).

**Comment by jerebtw** [comment]
Another little thing is a typo: https://github.com/pocketbase/pocketbase/blob/231ddc97919fd4374b19e7c35f04dde7a38f3077/daos/view.go#L130

Should be:
```
missing required id column (you can use `(ROW_NUMBER() OVER()) as id` if you don't have one)
```
